### ISSUE TITLE
Make the difference between `is` and `like` easier to understand.

### ISSUE THREAD

**Issue by schwern** [Test-More/Test2-Suite#63]
#56 and #62 bring up interesting questions about the behavior of `is` and `like`. The docs say `is` is the "strict checker" and `like` is the "relaxed checker". And I've been thinking of them this way. `is` is an equality check and `like` says we expect it to *contain* the selected things.

But upon reading the docs there's a lot more than that and some of them are quite surprising.

```
use Test2::Bundle::Extended;

use v5.10;

# This is "foo" =~ qr/foo/
like "foo", qr/foo/, "like regex";               # pass

# This is "foo" eq qr/foo/ ... or something?
is "foo",   qr/foo/,   "is regex";               # fails

# This uses the code ref to check 'foo'
like "foo", sub { $_ eq 'foo' }, 'like coderef'; # pass

# This checks if "foo" is the code ref?
is   "foo", sub { $_ eq 'foo' }, 'is coderef';   # fails

done_testing;
```

These `is` features are for extremely rare cases. They complicate understanding `is` vs `like`. The `like` ones are pretty useful, being able to check strings with regexes and evaluating complicated nested conditions with code refs. It sucks that you can only get them with `like`. You can't choose "strict checking but also let me use regexes and code".

I propose...

* Make `is` and `like` behave as similar as possible
  * Eliminate the code ref and regex and any other special cases
  * Replace them with `coderef` and `regex` DSLs (see below).
* `is` checks that the two sides are the same
* `like` checks that the LHS *contains* the RHS

The special case behavior of `is` is replaced by adding to the DSL. Then you can use it with `like` if you want.

    is $code, codref $some_code_ref;  # check that $code is $some_code_ref
    is $regex, regex qr{foo};  # check $regex is the same as qr{foo}

Yes, this does call into question what is really the difference between is and like.

Yes, I'm increasingly thinking of Test2 as a DSL rather than a collection of comparison functions.

**Comment by petdance** [comment]
> #56 and #62 bring up interesting questions about the behavior of is and like. 

See also #46.  I got bitten almost immediately.

**Comment by exodist** [comment]
This ticket is wrong in a couple ways.

You can use regex/codeblock in is, you just have to use the right DSL keywords:

    is($thing, match qr/foo/);
    like($thing, qr/foo/);

These 2 are the same. like simply assumes regexes get prefixed with match, and that coderefs are prefixed with validator. This behavior was chosen so that everyone's favorite construct ```like($thing, qr/foo/)``` would behave as expected with no changes.

Likewise you can use a direct compare of a regex or codeblock 'like'

    like(\@list, [exact_ref($array), exact_ref($regex)]);



Now, if you want to argue that the auto-convert behavior of 'like' is not ideal, then I propose we add a third check function to be the non-strict, but no auto-convert of regex/codeblock. I do not want to alter this behavior of like. That said the implementations of like and is are mostly shared code, and a subroutine ref is passed in for how to convert structures, so it would be trivial to write a third function that converts like is() does, but does not assume end().

Now what to call the third function? maybe contains()?

**Comment by schwern** [comment]
It's cool that the DSL words already exist. That's a side issue.

There's two things being mixed up here. First is the subtle differences between `is` and `like`, one auto-converts some things and one does not, and how that makes them even more difficult to understand than they already are. They're awesome, but they're not easy to wrap your head around. The second is whether auto-converting some things is a desirable behavior.

To the first, the subtle differences between `is` and `like` complicate understanding when to use one or the other. The differences are unexpected, difficult to predict, and rarely useful. I suspect we could count the number of people who really understand them on our fingers. If people don't understand how they work they're not going to use them. I see people using `like` for only `like $foo, qr/.../`. This calls into question whether we even need `like` at all.

To the second, the auto-convert behavior is `like` is the expected and most useful behavior. It's the behavior of `is` that is not ideal. I get now why you made that decision, you wanted `is` to be straight equality, but it's extraordinarily rare that you need to compare two regexes or two code refs. I suspect we could count the number of times we've ever wanted to check regex or code ref equality on our toes. Though I'm less sold on auto-converting code refs, that's going to be pretty rare too, and you probably don't want to surprise someone by executing code.

`contains` is a good name for what I'm driving at. It should match the behavior of `is` in all other ways. That would mean `is` should convert regexes into `match`. While that's a backwards compat thing, I really don't think anyone is using `is` to test for exact regexes or code refs... or even realize they can.

This is a big shift in thinking about Test2. We should talk about it in a better way than a series of Github issues.

**Comment by exodist** [comment]
@schwern I don't want to make too many changes to is/like, but I do not mind creating a new function that does things as you suggest. I am also fine with documenting that people should use the new one to avoid surprises. I don't want to say that is/like would be deprecated, but they would be documented as having behaviors that may surprise people, and should be used only when those behaviors are understood.

(I still want to change how is() handles hash/array to have implicit end)

Once I have the documentation ready it should be easy for you to come in and write 'contains' or whatever we decide to call it.

**Comment by exodist** [comment]
@schwern PR #68 would allow you to write contains(). PR #66 already addressed implicit end for is(), so once #68 is merged we can close this ticket and you can write contains().

**Comment by schwern** [comment]
Closing this down, we can discuss `contains` in #71.
